### ISSUE TITLE
Minio backend is killed by OOM under heavy load with errors

### ISSUE THREAD

**Issue by ordahan** [minio/minio#5994]
## Expected Behavior

Minio should keep a pretty stable memory footprint regardless of the user IO pattern, it should not leak / accumulate memory when presented with a lot of requests that subsequently fail due to bad server / client network.

## Current Behavior

After about a minute in which minio got a lot of invalid PUT requests, the kernel killed the process because it used too much memory (14GB):
```
May  9 18:35:54 REG-61081-V31MMOBS00-63695-minio-10 kernel: [ 3398.041942] Out of memory: Kill process 23098 (minio) score 934 or sacrifice child
May  9 18:35:54 REG-61081-V31MMOBS00-63695-minio-10 kernel: [ 3398.042908] Killed process 23098 (minio) total-vm:14506184kB, anon-rss:14365508kB, file-rss:0kB, shmem-rss:0kB
```

This was the repeated backtrace for the errors, taken from the syslog:
```
May  9 18:35:53 REG-61081-V31MMOBS00-63695-minio-10 minio: Error: unexpected EOF
May  9 18:35:53 REG-61081-V31MMOBS00-63695-minio-10 minio: 1: cmd/erasure-createfile.go:60:cmd.(*ErasureStorage).CreateFile()
May  9 18:35:53 REG-61081-V31MMOBS00-63695-minio-10 minio: 2: cmd/xl-v1-object.go:644:cmd.xlObjects.putObject()
May  9 18:35:53 REG-61081-V31MMOBS00-63695-minio-10 minio: 3: cmd/xl-v1-object.go:520:cmd.xlObjects.PutObject()
May  9 18:35:53 REG-61081-V31MMOBS00-63695-minio-10 minio: 4: cmd/xl-sets.go:560:cmd.(*xlSets).PutObject()
May  9 18:35:53 REG-61081-V31MMOBS00-63695-minio-10 minio: 5: cmd/object-handlers.go:669:cmd.(ObjectLayer).PutObject-fm()
May  9 18:35:53 REG-61081-V31MMOBS00-63695-minio-10 minio: 6: cmd/object-handlers.go:746:cmd.objectAPIHandlers.PutObjectHandler()
May  9 18:35:53 REG-61081-V31MMOBS00-63695-minio-10 minio: 7: cmd/api-router.go:77:cmd.(objectAPIHandlers).PutObjectHandler-fm()
May  9 18:35:53 REG-61081-V31MMOBS00-63695-minio-10 minio: 8: pkg/handlers/http-tracer.go:149:handlers.TraceReqHandlerFunc.func1()
May  9 18:35:53 REG-61081-V31MMOBS00-63695-minio-10 minio: 9: net/http/server.go:1947:http.HandlerFunc.ServeHTTP()
```
## Steps to Reproduce (for bugs)

1. Run a distributed minio cluster 
2. Send invalid HTTP requests with an incorrect body length (or no body at all, when the content-length is non zero)
3. Wait for the Minio server to get killed by the kernel (oom killer)

## Your Environment

* Version used (`minio version`):
Don't have the version in my logs, but it was whatever was served on https://dl.minio.io/server/minio/release/linux-amd64/minio at 2018-05-09 18:20:25 (UTC)
* Operating System and version (`uname -a`):
`May  9 18:20:19 REG-61081-V31MMOBS00-63695-minio-10 kernel: [    0.000000] Linux version 3.10.0-514.26.2.el7.x86_64 (builder@kbuilder.dev.centos.org) (gcc version 4.8.5 20150623 (Re
d Hat 4.8.5-11) (GCC) ) #1 SMP Tue Jul 4 15:04:05 UTC 2017`

**Comment by harshavardhana** [comment]
@ordahan could you share the reproducer script or tool?

**Comment by ordahan** [comment]
@harshavardhana I can't share my reproduction here as its not a simple script but its a test suite running a full blown storage system.

I think that by hacking the `mc` tool Its possible to recreate it, I believe the issue will get reproduced by uploading an object and closing the connection forcefully from the client side before the server completes the action (i.e almost 0 timeout in the client).

I peeked around in the `mc` code a bit - I think that by changing https://github.com/minio/mc/blob/c0b89a7aecb89174f3fda68f726eea2e3342b0aa/cmd/client-s3.go#L145 to a very short (lets say 1 msec) timeout, and then execute an `mc cp` of a lot of files (bombard the server with lots of operations) it will reproduce the issue.

**Comment by harshavardhana** [comment]
I tried reproducing with the following script which would throw 3000 requests in parallel

```bash
parallel -j 3000 ./oom.py ::: $(seq 100000)
```

```py
#!/usr/bin/env python

import urllib3
import certifi
import io

http = urllib3.PoolManager(
    timeout=1,
    maxsize=10,
    cert_reqs='CERT_REQUIRED',
    retries=urllib3.Retry(
        total=0,
    )
)

headers = {'Content-Length': 1024}
http.urlopen("PUT", "http://192.168.1.92:9001/testbucket/oom.txt",
             body=io.BytesIO(""),
             headers=headers)
```

I ran it overnight on a distributed setup. I am not able to crash the server, the memory usage also doesn't grow more than 0.3% to 2% of the overall RAM this system has. 
```
$ free -h
              total        used        free      shared  buff/cache   available
Mem:            62G        5.1G         56G         23M        835M         57G
```

I have a few more questions here 

- Were the requests that were sent signed already? or was it an anonymous PUT operation?
- How many requests per second were being sent, this would tell us more about how to reproduce it.
- Did you happen to load balance your requests to all minio servers in a distributed setup?
- What was the size of the cluster that this was tested on?

NOTE: I used latest release where we fixed a particular byte pool usage, there has been many improvements. Would you mind giving latest release and see if this can be reproduced??

**Comment by harshavardhana** [comment]
After some more testing, I can see why the memory usage goes up. The usage is an effect of how erasure coding needs to allocate a buffer even if the body is empty.  While this is not a leak - on a larger system like ours where we have 64GiB of memory I can see the memory usage with 1000 connections in parallel repeated for around 1000 times - each process goes to 10GiB. Golang GC is notorious for not cleaning up quickly as we want. 

```py
#!/usr/bin/env python

import urllib3
import certifi
import io
import random
import uuid

http = urllib3.PoolManager(
    timeout=10,
    maxsize=1024,
    cert_reqs='CERT_REQUIRED',
    retries=urllib3.Retry(
        total=0,
    )
)

name = uuid.uuid4().get_hex()
url =  "http://192.168.1.92:900{}/testbucket/{}.txt".format(random.randint(1, 4), name)
headers = {'Content-Length': 1024}
http.urlopen("PUT", url,
             body=io.BytesIO(""),
             headers=headers)
```

**Comment by harshavardhana** [comment]
The following PR merged into master fixes this to some extent almost a factor 10x, but we do still need a more comprehensive fix. Which basically rate limits the overall connections based on the RAM size. 

```
commit 05f96f3956d9d198aa690e655930daa60559607f (HEAD -> master, origin/master, origin/HEAD)
Author: Harshavardhana <harsha@minio.io>
Date:   Wed Jun 13 11:55:12 2018 -0700

    Avoid allocating larger buffer if the incoming body is small (#6035)
```

**Comment by ordahan** [comment]
Hey @harshavardhana sorry for the very late response, I was on vacation for the last 10 days.

After the PR is merged to master, when can I expect it in a release version that I can test? If not, where can I install the `master` version from? (currently my script downloads from https://dl.minio.io/server/minio/).

Regardless, I'll answer the above questions for documentation purposes:
* Requests were signed with AWSSignature4
* Request rate for a single host - I no longer have that info, but the concurrent connection number was about 64 to each host.
* Cluster was load balanced using round-robin DNS
* Cluster of 5 minio hosts

**Comment by ordahan** [comment]
Hey @harshavardhana let me know if this should go in a separate issue but I tested the fix (using the https://data.minio.io:10000/release-minio/staging/DEVELOPMENT.2018-06-20T08-07-36Z/linux-amd64/minio binary) and now instead of crashing on OOM, the Minio cluster stops being responsive, emitting 500 errors:
```
[REQUEST (objectAPIHandlers).PutObjectHandler-fm] [152948623.464018] [2018-06-20 12:17:14 +0300]
PUT /wekabucket-0/f1300856/d/d/99/2/29/536d-4903-a8f2-48a1c67d6fca
Host: minio.reg-68340-v31mmobs00-70778.wekalab.io
Accept: */*
Date: Wed, 20 Jun 2018 09:17:06 GMT
Content-Type: application/octet-stream
Authorization: AWS4-HMAC-SHA256 Credential=admin/20180620//s3/aws4_request,SignedHeaders=content-type;host;x-amz-content-sha256;x-amz-date,Signature=e90a7029ce6d101ee42fc5a083a6fb72
e616d47c26982dd77b7cbc925fc64747
X-Amz-Date: 20180620T091706Z
Content-Length: 524326
Expect: 100-continue
User-Agent: APN/1.0 WEKA/1.0 WEKASCALE/V3.0
X-Amz-Content-Sha256: 01d36280dc37ef533c9d78baad5e2c7d91fa6363efc79675d53334dbb8f44765

<BODY>

[RESPONSE] [152948623.464018] [2018-06-20 12:30:32 +0300]
500 Internal Server Error
Server: Minio/DEVELOPMENT.2018-06-20T08-07-36Z (linux; amd64)
Accept-Ranges: bytes
Content-Type: application/xml
Vary: Origin
X-Xss-Protection: 1; mode=block
Content-Security-Policy: block-all-mixed-content
X-Amz-Request-Id: 1539D45A37D0631F

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>InternalError</Code><Message>We encountered an internal error, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/wekabucket-0/f1300856/d/d/99/2/
29/536d-4903-a8f2-48a1c67d6fca</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
```
and I dug a bit in the syslog, I got these sort of errors:
```
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: API: PutObject(bucket=wekabucket-0, object=f1300856/d/d/99/2/29/536d-4903-a8f2-48a1c67d6fca)
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: Time: 12:21:16 IDT 06/20/2018
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: RemoteHost: 172.31.18.80:57862
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: UserAgent: APN/1.0 WEKA/1.0 WEKASCALE/V3.0
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: Error: read tcp 172.31.18.181:80->172.31.18.80:57862: read: connection reset by peer
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: 1: cmd/erasure-createfile.go:60:cmd.(*ErasureStorage).CreateFile()
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: 2: cmd/xl-v1-object.go:656:cmd.xlObjects.putObject()
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: 3: cmd/xl-v1-object.go:523:cmd.xlObjects.PutObject()
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: 4: cmd/xl-sets.go:564:cmd.(*xlSets).PutObject()
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: 5: cmd/object-handlers.go:731:cmd.(ObjectLayer).PutObject-fm()
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: 6: cmd/object-handlers.go:808:cmd.objectAPIHandlers.PutObjectHandler()
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: 7: cmd/api-router.go:79:cmd.(objectAPIHandlers).PutObjectHandler-fm()
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: 8: pkg/handlers/http-tracer.go:149:handlers.TraceReqHandlerFunc.func1()
Jun 20 12:21:16 REG-68340-V31MMOBS00-70778-minio-11 minio: 9: net/http/server.go:1947:http.HandlerFunc.ServeHTTP()

Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: API: PutObject(bucket=wekabucket-3, object=f1300856/d/d/155/1/25/536d-4903-a8f2-48a1c67d6fca)
Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: Time: 12:31:01 IDT 06/20/2018
Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: RemoteHost: 172.31.18.95:58210
Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: UserAgent: APN/1.0 WEKA/1.0 WEKASCALE/V3.0
Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: Error: unexpected EOF
Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: 1: cmd/erasure-createfile.go:60:cmd.(*ErasureStorage).CreateFile()
Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: 2: cmd/xl-v1-object.go:656:cmd.xlObjects.putObject()
Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: 3: cmd/xl-v1-object.go:523:cmd.xlObjects.PutObject()
Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: 4: cmd/xl-sets.go:564:cmd.(*xlSets).PutObject()
Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: 5: cmd/object-handlers.go:731:cmd.(ObjectLayer).PutObject-fm()
Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: 6: cmd/object-handlers.go:808:cmd.objectAPIHandlers.PutObjectHandler()
Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: 7: cmd/api-router.go:79:cmd.(objectAPIHandlers).PutObjectHandler-fm()
Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: 8: pkg/handlers/http-tracer.go:149:handlers.TraceReqHandlerFunc.func1()
Jun 20 12:31:01 REG-68340-V31MMOBS00-70778-minio-11 minio: 9: net/http/server.go:1947:http.HandlerFunc.ServeHTTP()

Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: API: PutObject(bucket=wekabucket-3, object=f1300856/d/d/110/1/25/536d-4903-a8f2-48a1c67d6fca)
Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: Time: 12:30:40 IDT 06/20/2018
Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: RemoteHost: 172.31.18.230:53564
Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: UserAgent: APN/1.0 WEKA/1.0 WEKASCALE/V3.0
Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: Error: unexpected EOF
Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: 1: cmd/erasure-createfile.go:60:cmd.(*ErasureStorage).CreateFile()
Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: 2: cmd/xl-v1-object.go:656:cmd.xlObjects.putObject()
Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: 3: cmd/xl-v1-object.go:523:cmd.xlObjects.PutObject()
Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: 4: cmd/xl-sets.go:564:cmd.(*xlSets).PutObject()
Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: 5: cmd/object-handlers.go:731:cmd.(ObjectLayer).PutObject-fm()
Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: 6: cmd/object-handlers.go:808:cmd.objectAPIHandlers.PutObjectHandler()
Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: 7: cmd/api-router.go:79:cmd.(objectAPIHandlers).PutObjectHandler-fm()
Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: 8: pkg/handlers/http-tracer.go:149:handlers.TraceReqHandlerFunc.func1()
Jun 20 12:30:40 REG-68340-V31MMOBS00-70778-minio-11 minio: 9: net/http/server.go:1947:http.HandlerFunc.ServeHTTP()

Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: API: PutObject(bucket=wekabucket-3, object=f1300856/d/d/117/1/9/536d-4903-a8f2-48a1c67d6fca)
Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: Time: 12:30:36 IDT 06/20/2018
Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: RemoteHost: 172.31.18.95:58151
Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: UserAgent: APN/1.0 WEKA/1.0 WEKASCALE/V3.0
Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: Error: unexpected EOF
Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: 1: cmd/erasure-createfile.go:60:cmd.(*ErasureStorage).CreateFile()
Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: 2: cmd/xl-v1-object.go:656:cmd.xlObjects.putObject()
Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: 3: cmd/xl-v1-object.go:523:cmd.xlObjects.PutObject()
Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: 4: cmd/xl-sets.go:564:cmd.(*xlSets).PutObject()
Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: 5: cmd/object-handlers.go:731:cmd.(ObjectLayer).PutObject-fm()
Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: 6: cmd/object-handlers.go:808:cmd.objectAPIHandlers.PutObjectHandler()
Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: 7: cmd/api-router.go:79:cmd.(objectAPIHandlers).PutObjectHandler-fm()
Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: 8: pkg/handlers/http-tracer.go:149:handlers.TraceReqHandlerFunc.func1()
Jun 20 12:30:36 REG-68340-V31MMOBS00-70778-minio-11 minio: 9: net/http/server.go:1947:http.HandlerFunc.ServeHTTP()
```

Its possible that the trigger for these errors is an aggressive network stack on our side that creates lots of connections, reuses connections etc 

What more info do you guys need in order to analyze the issue?

**Comment by harshavardhana** [comment]
500 errors are what would be expected since the body is smaller than the content-length and you are correctly looking at unexpected eof. 

What does the application expect as an error ?

**Comment by harshavardhana** [comment]
@ordahan we have made a release with the fix would you mind testing it and let us know if you face further issues. I didn't a reply to my previous message though reposting it here 

> 500 errors are what would be expected since the body is smaller than the content-length and you are correctly looking at unexpected eof.
> What does the application expect as an error ?

**Comment by ordahan** [comment]
@harshavardhana i'm testing the latest release now
500 errors are definitely in place for these content-length issues, I mistakenly thought that requests with a correct content-length were being rejected as well.

**Comment by harshavardhana** [comment]
@ordahan okay, in that case should we closing this issue ?

**Comment by divanikus** [comment]
I'm keep having the same issue on the latest version. Is there any config option for that?

**Comment by harshavardhana** [comment]
> I'm keep having the same issue on the latest version. Is there any config option for that?

Feel free to open an issue @divanikus with some more information.
